ddl/column.go

```diff
@@ -222,14 +222,249 @@ func onAddColumn(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error)
 
 		// Finish this job.
 		job.FinishTableJob(model.JobStateDone, model.StatePublic, ver, tblInfo)
-		asyncNotifyEvent(d, &util.Event{Tp: model.ActionAddColumn, TableInfo: tblInfo, ColumnInfo: columnInfo})
+		asyncNotifyEvent(d, &util.Event{Tp: model.ActionAddColumn, TableInfo: tblInfo, ColumnInfos: []*model.ColumnInfo{columnInfo}})
 	default:
 		err = ErrInvalidDDLState.GenWithStackByArgs("column", columnInfo.State)
 	}
 
 	return ver, errors.Trace(err)
 }
 
+func checkAddAndCreateColumnInfos(t *meta.Meta, job *model.Job) (*model.TableInfo, []*model.ColumnInfo, []*ast.ColumnPosition, []int, error) {
+	schemaID := job.SchemaID
+	tblInfo, err := getTableInfoAndCancelFaultJob(t, job, schemaID)
+	if err != nil {
+		return nil, nil, nil, nil, errors.Trace(err)
+	}
+	columns := []*model.ColumnInfo{}
+	positions := []*ast.ColumnPosition{}
+	offsets := []int{}
+	err = job.DecodeArgs(&columns, &positions, &offsets)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return nil, nil, nil, nil, errors.Trace(err)
+	}
+	// Should initialize positions[i] when it is nil, otherwise the createColumnInfo function will panic.
+	for i := range positions {
+		if positions[i] == nil {
+			positions[i] = &ast.ColumnPosition{}
+		}
+	}
```

Comment:
Could we put line248-252 to `createColumnInfo`? Then we needn't check `if positions[i] == nil`.

Reply:
Hi, we can't do this because `positions` has used in :
```
	job.Args = []interface{}{columnInfos, positions, offsets}
	return tblInfo, columnInfos, positions, offsets, nil
```

Reply:
I don't understand, we initialized `positions` on line240. If the first time call `checkAddAndCreateColumnInfos`, we will execute `createColumnInfo`, then goto line281. If it is not the first time to call `checkAddAndCreateColumnInfos`， the values in `positions` need not to initialize.

Reply:
I can't change items of the array `positions` in for range loop. And I add `pos` as return value of function `createColumnInfo`. Is this okay？

Reply:
But it seems that `col` is used as `for range loop`. `positions` use the format like `positions[i]`, the value can be changed. 
It's ok to use away like `pos`.